### PR TITLE
NEW Adding reset and remove buttons to managing MFA methods

### PR DESCRIPTION
**Pull request by ScopeyNZ**
Adds "remove" and "reset" buttons to registered methods against yourself in the CMS.

### CODE REVIEW

**Comment by NightJar** [comment]
Will this not also address #35 ?

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.scss [hunk 1]
```diff
@@ -1,9 +1,40 @@
...
+    }
+
+    & + & {
+      margin-left: $input-padding-x-lg;
+    }
+  }
+}
+
+.registered-mfa-method-list-field-register-modal {
+  height: initial;
```
Comment (ScopeyNZ) [inline client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.scss]: @robbieaverill This is a fix for that weird overflow issue you pointed out to me on Slack.

File: client/src/bundles/bundle-cms.scss [hunk 2]
```diff
@@ -11,7 +12,12 @@
 @import '../styles/mfa-variables';
 
 // Component styles
+@import '../components/LoadingIndicator';
```
Comment (robbieaverill) [inline client/src/bundles/bundle-cms.scss]:
```suggestion
```

It's there below

File: client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js [hunk 3]
```diff
@@ -28,7 +28,26 @@ class AccountResetUI extends Component {
    * Sends a reset request to the provided endpoint, and updates the component's state based on
    * the contents of the response.
    */
-  handleSendReset() {
+  async handleSendReset() {
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js]: Do we need any kind of polyfill to support async methods?

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js [hunk 4]
```diff
@@ -1,129 +1,429 @@
...
+import Config from 'lib/Config'; // eslint-disable-line
+import confirm from '@silverstripe/reactstrap-confirm';
 
 const fallbacks = require('../../../../lang/src/en.json');
 
 class RegisteredMFAMethodListField extends Component {
   constructor(props) {
     super(props);
 
+    const { initialAvailableMethods, registeredMethods } = props;
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: Adding 444 lines to a component seems wrong to me. Any way you could split it up?

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js [hunk 5]
```diff
@@ -1,129 +1,429 @@
...
-  handleToggleModal() {
+  handleToggleModal(skipReset = false) {
     const { modalOpen } = this.state;
 
     this.setState({
       modalOpen: !modalOpen,
     });
 
-    if (!modalOpen) {
+    if (!modalOpen && skipReset !== true) {
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: Can we just say `!skipReset`? It seems like you're implementing your own kind of strict typing logic here, but not sure it's really that necessary

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js [hunk 6]
```diff
@@ -1,129 +1,429 @@
...
+   */
+  handleRemoveMethod(method) {
+    const { onAddAvailableMethod } = this.props;
+
+    this.setState(state => ({
+      registeredMethods: state.registeredMethods.filter(
+        candidate => candidate.urlSegment !== method.urlSegment
+      ),
+    }));
+    onAddAvailableMethod(method);
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: There's no guarantee that setState is immediate, you can move this into a callback for the second argument of `setState` to fix this

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js [hunk 7]
```diff
@@ -1,129 +1,429 @@
...
+   *
+   * @param {object} method
+   */
+  handleResetMethod(method) {
+    const registerableMethod = this.props.allAvailableMethods.find(
+      candidate => candidate.urlSegment === method.urlSegment
+    );
+
+    if (!registerableMethod) {
+      throw Error(`Cannot register the method given: ${method.name} (${method.urlSegment}).`);
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: Are these for developers or for the UI? If the latter, should we make them translatable?

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.scss [hunk 8]
```diff
@@ -1,9 +1,40 @@
 .registered-mfa-method-list-field {
   padding: $input-padding-y 0;
+
+  &__button, &__no-methods {
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.scss]: I think this should have a line drop after the comma, surprised sass lint doesn't pick that up

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.scss [hunk 9]
```diff
@@ -1,9 +1,40 @@
...
+    text-decoration: none;
+    color: $link-color;
+    cursor: pointer;
+
+    &:hover {
+      color: $link-hover-color;
+      text-decoration: underline;
+    }
+
+    & + & {
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.scss]: Can you comment what this does? I always get confused when I read these

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.scss (hunk 1 above)
Comment (robbieaverill) [reply client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.scss]: Fancy adding a comment to say that's what it's for?

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js [hunk 10]
```diff
@@ -160,16 +449,21 @@ RegisteredMFAMethodListField.propTypes = {
   backupMethod: methodShape,
   defaultMethod: methodShape,
   readOnly: PropTypes.bool,
-  availableMethods: PropTypes.arrayOf(methodShape),
+  initialAvailableMethods: PropTypes.arrayOf(methodShape),
```
Comment (robbieaverill) [inline client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: Needs `allAvailableMethods` added as well

File: src/Controller/AdminRegistrationController.php [hunk 11]
```diff
@@ -77,7 +105,65 @@ public function finishRegistration(HTTPRequest $request): HTTPResponse
...
+                ['errors' => [_t(__CLASS__ . '.CSRF_FAILURE', 'Request timed out, please try again')]],
+                400
+            );
+        }
+
+        // Get the specified method
+        $methodRegistry = MethodRegistry::singleton();
+        $method = $request->param('Method');
+
+        if (!$method || !($method = $methodRegistry->getMethodByURLSegment($request->param('Method')))) {
```
Comment (robbieaverill) [inline src/Controller/AdminRegistrationController.php]: Presumably the second half of this condition would return false with a null argument anyway, so we need the first bit?

Comment (ScopeyNZ) [reply src/Controller/AdminRegistrationController.php]: Strict typing requires that we don't pass in `null` - so we need the check. I can probably make this a little nicer by not re-declaring the same named variable though so I'll do that.

File: client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js (hunk 3 above)
Comment (ScopeyNZ) [reply client/src/components/FormField/RegisteredMFAMethodListField/AccountResetUI.js]: AFAICT it's supported by babel given our browser target configuration.

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js (hunk 5 above)
Comment (ScopeyNZ) [reply client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: Yeah I don't know why I did that.

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js (hunk 6 above)
Comment (ScopeyNZ) [reply client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: I am using a callback for `setState`? I'm confused.

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js (hunk 7 above)
Comment (ScopeyNZ) [reply client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: It's probably a redundant error. The UI should make this impossible - meaning it's an error for devs.

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js (hunk 6 above)
Comment (robbieaverill) [reply client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: `onAddAvailableMethod` should be the callback right? Its not right now

File: client/src/components/FormField/RegisteredMFAMethodListField/MethodListItem.js [hunk 12]
```diff
@@ -11,20 +15,118 @@ const fallbacks = require('../../../../lang/src/en.json');
...
+
+    const controls = [];
+    const { ss: { i18n } } = window;
+
+    const buildControl = (action, text) => ((
+      <button
+        className="registered-method-list-item__control"
+        key={text}
+        onClick={event => {
+          event.preventDefault();
```
[the remark this answers is not in the document]
Comment (ScopeyNZ) [reply client/src/components/FormField/RegisteredMFAMethodListField/MethodListItem.js]: Nice. Didn't realise that had the same effect. Probably doesn't hurt to do both?

[the remark this answers is not in the document]
Comment (Cheddam) [reply client/src/components/FormField/RegisteredMFAMethodListField/MethodListItem.js]: `<button type="button">` should be sufficiently inert, but up to you 👍

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js (hunk 6 above)
Comment (ScopeyNZ) [reply client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: Ok I've refactored this a bit now anyway.

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js (hunk 4 above)
Comment (ScopeyNZ) [reply client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: Ok I've done a bit of splitting up now. Let me know if that suffices.

File: client/src/state/mfaAdministration/reducer.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+      return {
+        ...state,
+        registeredMethods: payload.methods,
+      };
+    }
+
+    default:
+      return state;
+  }
+}
```
Comment (robbieaverill) [inline client/src/state/mfaAdministration/reducer.js]: Some doc blocks on each action here might be useful to describe what's going on

**Pull request merged**